### DISCUSSION TITLE
How do I consume this package as a Nix derivation?

### DISCUSSION THREAD

**Discussion by mboyea** [Infinidoge/nix-minecraft#106] [Q&A] [answered]
Hello!

I would like to create a Minecraft server in a Docker image using Nix. I have already created the necessary build and deployment scripts, and now it is time for me to pack a Minecraft server into my image. My public repository for this purpose is at [github.com/mboyea/nixflymc](https://github.com/mboyea/nixflymc), where the Minecraft server would go inside `src/server.nix`.

IIUC existing support for the `Infinidoge/nix-minecraft` package seems to expect for it to be added to a NixOS configuration, and run as a `systemd-service`. While it is almost certainly possible for me to create a NixOS Docker image, I would prefer to consume this package as a Nix derivation instead. That way, I could produce a distro-less Docker image for the server (which is much more lean).

What options do I have?
Thank you for your time!

**Comment by Infinidoge** [comment]
`nix-minecraft` is a flake that exports both a module and a package set. If your goal is to consume a package, then you just get a package from the `packages` or `legacyPackages` package sets (documented in the readme). The flake doesn't export a systemd service derivation, however.

If your case, you'd add
```nix
inputs.nix-minecraft.url = "github:Infinidoge/nix-minecraft";
```
And in your flake, you'd use `nix-minecraft` as the input.
Notably, the package outputs are already run scripts, so you shouldn't need to add your own. `nix-minecraft.packages.x86_64-linux.vanilla-server` would work for the latest vanilla server, for example.

**Reply by mboyea** [reply]
Thank you for your response.
I understand how to pull your package into the flake.
It makes sense to me that I could run it directly as you described.
Please bear with me regarding my following questions, as I am relatively new to Nix.

### How do I make a call to this package outside `flake.nix`?

I want to have the file `flake.nix` serve only as a package manager (no business logic of the app should be included), like:
```nix
{
  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixos-24.05";
    utils.url = "github:numtide/flake-utils";
    nix-minecraft.url = "github:Infinidoge/nix-minecraft";
  };
  outputs = { self, nixpkgs, utils, nix-minecraft, ... }: let
    name = "nixflymc";
    version = "0.0.0";
  in utils.lib.eachDefaultSystem (
    system: let
      pkgs = import nixpkgs { inherit system; };
    in rec {
      packages = {
        server = pkgs.callPackage ./src/server.nix {
          inherit name version system nix-minecraft;
        };
        default = packages.server;
      };
      apps = {
        server = utils.lib.mkApp { drv = packages.server; };
        default = apps.server;
      };
    }
  );
}
```
And then the file `src/server.nix` will contain a call to `nix-minecraft` something like:
```nix
{ stdenv, name, version, system, nix-minecraft } : stdenv.mkDerivation {
  pname = "${name}-server";
  inherit version;
  name = "${pname}-${version}";
  src = nix-minecraft.packages.${system}.vanilla-server;
  # ...?
}
```
This file does not build because the derivation doesn't actually do anything.

### How do I wrap the call to this package with a configuration?

I still don't see how to pass *configuration* into the server. Existing instructions describe how to place options into the `configuration.nix` file, rather than exposing them to the package with arguments.

I'd like to contain the configuration within the `src/server.nix` file.

### How does this affect server configuration?

I'm not sure how boot-specific options like `services.minecraft-servers` would be affected.
`servers.<name>` says, "This family of options govern individual servers, which will be created on boot."
`servers.<name>.autoStart` says, "Whether to start this server on boot. If set to false, can still be started with `systemctl start minecraft-server-servername`. Requires the server to be enabled."

If this package isn't installed as a part of NixOS, what is the behavior of these configuration options?

**Reply by mboyea** [reply]
Trying to call the package with no configuration results in the following output:

```sh
[15:27:47] [ServerMain/WARN]: Failed to load eula.txt
[15:27:47] [ServerMain/INFO]: You need to agree to the EULA in order to run the server. Go to eula.txt for more info.
```

Because of course, we need to set `eula = true`. How do I get that configuration into the package?

After taking a look at the source code and some more Nix documentation, it looks like the NixOS options are declared by [nix-minecraft/modules/minecraft-servers.nix](https://github.com/Infinidoge/nix-minecraft/blob/master/modules/minecraft-servers.nix). Unless there is a way to consume NixOS modules inside my own Nix package (`src/server.nix`), it seems I'd have to re-implement all of the options I need offered by `minecraft-servers.nix` inside `src/server.nix`.

**Reply by Infinidoge** [reply, marked as answer]
It is possible but annoying on both accounts. An additional problem is that the `minecraft-servers` module is built around hosting multiple servers and having the ability to set other NixOS options which don't apply.

Fundamentally, if you are making this a docker container, you will need to rethink and reimplement how it is configured mostly from the ground up. The NixOS module is built around being run in NixOS, after all, and not a minimum docker container.

The packages don't handle configuration at all. That is entirely handled through systemd manipulating files. The EULA is handled the same way.

**Comment by mboyea** [comment]
## Conclusions

In the end, I decided to mimic the design pattern in [the nix-minecraft package derivation.nix code](https://github.com/Infinidoge/nix-minecraft/blob/master/pkgs/vanilla-servers/derivation.nix). The solution wraps the target package (`nix-minecraft`) in a new Nix derivation, within which you can patch in any modifications you desire.

In `src/server.nix`:

```nix
{ stdenv, name, version, system, nix-minecraft } : stdenv.mkDerivation rec {
  pname = "${name}-server";
  inherit version;
  src = nix-minecraft.packages.${system}.vanilla-server;
  installPhase = ''
    mkdir -p $out/bin $out/lib
    cp -r -v $src $out/lib/minecraft-server

    echo "echo 'eula=TRUE' > eula.txt" > $out/bin/${pname}

    cat >> $out/bin/${pname} << EOF
      $out/lib/minecraft-server/bin/minecraft-server
    EOF
    chmod +x $out/bin/${pname}
  '';
}
```
I don't love that the bash script which starts the server also automatically overwrites & accepts the EULA. This gets me started, but in the future I hope to figure out a better pattern through which I can configure the server. I am sure there is a more "Nix" way to patch in the configuration.

I've marked this Discussion as answered. For anyone interested, I will keep my current solutions public at [mboyea/nixflymc](https://github.com/mboyea/nixflymc/).

**Reply by Infinidoge** [reply]
Please don't do this. Patching the script is the wrong way to handle it. Instead, have your docker container create the actual files in the filesystem like the systemd units do.

**Reply by mboyea** [reply]
Yeah, like I said, I know using a bash script isn't the Nix way to do this.

However, the server itself needs to be a self-contained Nix package that works on its own. I don't think that the right solution is to bootstrap the files in with the Docker container. `nix run .#server` and `nix run .#serverContainer` should functionally operate the same way.

I'm trying to find a better way than bash file manipulation, like I said. I'm sure I can find the nix way to make it happen in `src/server.nix`!

It's just that there aren't a lot of examples on GitHub, and the documentation is very brief and disjointed. So it will take time a bit more to work out myself. Someone can check `src/server.nix` in [my repository](https://github.com/mboyea/nixflymc) if they want to see if I found a better pattern.
